dowhy/causal_identifier/backdoor_identifier.py

```diff
@@ -36,119 +22,160 @@ class CausalIdentifier:
+                    CausalIdentifierEstimandType.NONPARAMETRIC_CDE,
+                    CausalIdentifierEstimandType.NONPARAMETRIC_NDE,
+                    CausalIdentifierEstimandType.NONPARAMETRIC_NIE,
                 )
             )
 
-    def identify_ate_effect(self, optimize_backdoor, costs=None, conditional_node_names=None):
+    def identify_ate_effect(
+        self, graph: CausalGraph, treatment_name: List[str], outcome_name: str, conditional_node_names: List[str] = None
+    ):
         estimands_dict = {}
         mediation_first_stage_confounders = None
         mediation_second_stage_confounders = None
         ### 1. BACKDOOR IDENTIFICATION
         # Pick algorithm to compute backdoor sets according to method chosen
-        if self.method_name not in CausalIdentifier.EFFICIENT_METHODS:
+        if self.backdoor_adjustment not in BackdoorIdentifier.EFFICIENT_METHODS:
             # First, checking if there are any valid backdoor adjustment sets
-            if optimize_backdoor == False:
-                backdoor_sets = self.identify_backdoor(self.treatment_name, self.outcome_name)
+            if self.optimize_backdoor == False:
+                backdoor_sets = self.identify_backdoor(graph, treatment_name, outcome_name)
             else:
-                from dowhy.causal_identifiers.backdoor import Backdoor
+                from dowhy.causal_identifier.backdoor import Backdoor
 
-                path = Backdoor(self._graph._graph, self.treatment_name, self.outcome_name)
+                path = Backdoor(graph._graph, treatment_name, outcome_name)
                 backdoor_sets = path.get_backdoor_vars()
-        elif self.method_name in CausalIdentifier.EFFICIENT_METHODS:
-            backdoor_sets = self.identify_efficient_backdoor(costs=costs, conditional_node_names=conditional_node_names)
+        elif self.backdoor_adjustment in BackdoorIdentifier.EFFICIENT_METHODS:
+            backdoor_sets = self.identify_efficient_backdoor(graph, conditional_node_names=conditional_node_names)
         estimands_dict, backdoor_variables_dict = self.build_backdoor_estimands_dict(
-            self.treatment_name, self.outcome_name, backdoor_sets, estimands_dict
+            graph, treatment_name, outcome_name, backdoor_sets, estimands_dict
         )
         # Setting default "backdoor" identification adjustment set
-        default_backdoor_id = self.get_default_backdoor_set_id(backdoor_variables_dict)
+        default_backdoor_id = self.get_default_backdoor_set_id(
+            graph, treatment_name, outcome_name, backdoor_variables_dict
+        )
         if len(backdoor_variables_dict) > 0:
             estimands_dict["backdoor"] = estimands_dict.get(str(default_backdoor_id), None)
             backdoor_variables_dict["backdoor"] = backdoor_variables_dict.get(str(default_backdoor_id), None)
         else:
             estimands_dict["backdoor"] = None
         ### 2. INSTRUMENTAL VARIABLE IDENTIFICATION
```

Comment:
Is it odd that the backdoor_identifier class also has logic for IV estimation and frontdoor estimation?  Is there a better name for this class, or should we make an issue to refactor and separate this out?  FYI @amit-sharma

Reply:
I'm open to refactor this, maybe we could go with something generic like `DefaultIdentifier`.

In the future if we take the refactor approach to separate this then maybe we could also simplify the `IdentifiedEstimand` object, on the other hand users would need to call `identify_effect` 3 times with very similar parameters (only the `method` would change) to get the current result.

Reply:
Yes, let's resolve this by renaming to DefaultIdentifier right now.

When we refactor, I think we want to also update the Identify_effect function so that, if you do not specify a method (or give an array of methods) that the identify_effect method just calls a pre-determined set of default methods and returns an array of identified estimands.

Reply:
One (unsolicited, sorry) comment on this one. I think having IV there together with BD and FD is not ideal, mainly because, at least AFAIK, the graphical criteria for finding conditional IV sets do not suffice to identify the ATE, unless one is willing to make additional homogeneity type assumptions. See [this](https://arxiv.org/pdf/1611.09925.pdf) for example.

Reply:
I'll mark this conversation as resolved, now that this class is called AutoIdentifier.  

@esmucler, I agree that we need to track the additional, non-graphical, criteria for IV.  We are planning to add more explicit tracking of all assumptions so that we can either warn users and/or more automatically trigger the appropriate refutations, validations, and sensitivity analyses.  I think this would be a first candidate for being tracked when we have that set up.